tests/v2/validation/provisioning/rke2/agent_customization_test.go

```diff
@@ -0,0 +1,244 @@
+	}
+
+	for _, tt := range tests {
+		subSession := r.session.NewSession()
+		defer subSession.Cleanup()
+
+		client, err := tt.client.WithSession(subSession)
+		require.NoError(r.T(), err)
+		r.provisioningConfig.MachinePools = tt.machinePools
+
+		if tt.agent == "fleet-agent" {
+			r.provisioningConfig.FleetAgent = &agentCustomization
+			r.provisioningConfig.ClusterAgent = nil
+		}
+
+		if tt.agent == "cluster-agent" {
+			r.provisioningConfig.ClusterAgent = &agentCustomization
+			r.provisioningConfig.FleetAgent = nil
+		}
+
+		permutations.RunTestPermutations(&r.Suite, tt.name, client, r.provisioningConfig, permutations.RKE2ProvisionCluster, nil, nil)
+	}
+}
+
+func (r *RKE2AgentCustomizationTestSuite) TestProvisioningRKE2NegativeClusterAgentCustomization() {
+	productionPool := []provisioninginput.MachinePools{
+		{
+			NodeRoles: machinepools.NodeRoles{
+				ControlPlane: true,
+				Etcd:         false,
+				Worker:       false,
+				Quantity:     2,
+			},
+		},
+		{
+			NodeRoles: machinepools.NodeRoles{
+				ControlPlane: false,
+				Etcd:         true,
+				Worker:       false,
+				Quantity:     3,
+			},
+		},
+		{
+			NodeRoles: machinepools.NodeRoles{
+				ControlPlane: false,
+				Etcd:         false,
+				Worker:       true,
+				Quantity:     2,
+			},
+		},
+	}
+
+	agentCustomization := management.AgentDeploymentCustomization{
+		AppendTolerations: []management.Toleration{
+			{
+				Key:   "BadLabel",
+				Value: "123\"[];'{}-+=",
+			},
+		},
+		OverrideAffinity:             &management.Affinity{},
+		OverrideResourceRequirements: &management.ResourceRequirements{},
+	}
+
+	customAgents := []string{"fleet-agent", "cluster-agent"}
+	tests := []struct {
+		name         string
+		machinePools []provisioninginput.MachinePools
+		client       *rancher.Client
+		agent        string
+	}{
+		{"Bad Custom Fleet Agent - Standard User", productionPool, r.standardUserClient, customAgents[0]},
+		{"Bad Custom Cluster Agent - Standard User", productionPool, r.standardUserClient, customAgents[1]},
+	}
+
+	for _, tt := range tests {
+		subSession := r.session.NewSession()
+		defer subSession.Cleanup()
+
+		client, err := tt.client.WithSession(subSession)
+		require.NoError(r.T(), err)
+		r.provisioningConfig.MachinePools = tt.machinePools
+
+		if tt.agent == "fleet-agent" {
+			r.provisioningConfig.FleetAgent = &agentCustomization
+			r.provisioningConfig.ClusterAgent = nil
+		}
+
+		if tt.agent == "cluster-agent" {
+			r.provisioningConfig.ClusterAgent = &agentCustomization
+			r.provisioningConfig.FleetAgent = nil
+		}
+
+		rke2Provider, _, _, kubeVersions := permutations.GetClusterProvider(permutations.RKE2ProvisionCluster, r.provisioningConfig.Providers[0], r.provisioningConfig)
+		testClusterConfig := clusters.ConvertConfigToClusterConfig(r.provisioningConfig)
+		testClusterConfig.KubernetesVersion = kubeVersions[0]
+
+		_, err = provisioning.CreateProvisioningCluster(client, *rke2Provider, testClusterConfig, nil)
```

Comment:
Maybe we can do something similar to what we do for checking the cluster type in the etcd snapshot test suite. In other words, rather than throwing a failure, if the exit code comes to a failure, then we mark that as a "successful" test.

Maybe we can have an offline discussion on this as I think this is a good opportunity to establish a best practices for non-happy path tests.

Reply:
spoke with Israel offline and I think for now what is here is good for now, but definitely open to discussing in a call if you'd like.